src/executor.cc

```diff
@@ -32,65 +32,73 @@ Result Executor::Execute(Engine* engine,
                          const ShaderMap& shader_map) {
   engine->SetEngineData(script->GetEngineData());
 
-  // Process Shader nodes
-  PipelineType pipeline_type = PipelineType::kGraphics;
-  for (const auto& shader : script->GetShaders()) {
+  if (script->GetPipelines().empty())
+    return Result("no pipelines defined");
+
+  // TODO(jaebaek): Support multiple pipelines.
+  if (script->GetPipelines().size() > 1)
+    return Result("only single pipeline supported currently");
+
+  auto* pipeline = script->GetPipelines()[0].get();
+
+  // Process Shaders
+  for (const auto& shader_info : pipeline->GetShaders()) {
     ShaderCompiler sc(script->GetSpvTargetEnv());
 
     Result r;
     std::vector<uint32_t> data;
-    std::tie(r, data) = sc.Compile(shader.get(), shader_map);
+    std::tie(r, data) = sc.Compile(shader_info.GetShader(), shader_map);
     if (!r.IsSuccess())
       return r;
 
-    r = engine->SetShader(shader->GetType(), data);
+    r = engine->SetShader(shader_info.GetShaderType(), data);
     if (!r.IsSuccess())
       return r;
-
-    if (shader->GetType() == kShaderTypeCompute)
-      pipeline_type = PipelineType::kCompute;
   }
 
   // Handle Image and Depth buffers early so they are available when we call
   // the CreatePipeline method.
-  for (const auto& buf : script->GetBuffers()) {
-    // Image and depth are handled earlier. They will be moved to the pipeline
-    // object when it exists.
-    if (buf->GetBufferType() != BufferType::kColor &&
-        buf->GetBufferType() != BufferType::kDepth) {
-      continue;
-    }
+  for (const auto& info : pipeline->GetColorAttachments()) {
+    Result r = engine->SetBuffer(info.type, static_cast<uint8_t>(info.location),
+        info.buffer->AsFormatBuffer()->GetFormat(), info.buffer->GetData());
+    if (!r.IsSuccess())
+      return r;
+  }
 
-    Result r = engine->SetBuffer(
-        buf->GetBufferType(), buf->GetLocation(),
-        buf->IsFormatBuffer() ? buf->AsFormatBuffer()->GetFormat() : Format(),
-        buf->GetData());
+  if (pipeline->GetDepthBuffer().buffer) {
+    const auto& info = pipeline->GetDepthBuffer();
```

Comment:
Question: What happens if the first pipeline has a depth buffer but the second one doesn't ?  When is the engine supposed to reset its state?

Reply:
VkScript only has one pipeline. There is no concept so I create a fake one. So, for vkscript this will work fine.

For AmberScript we need to update the engine code to handle multiple pipelines so each setBuffer and setShader has to work per-pipeline not globally. I'm leaving that for a future enhancement as, I believe, this is the minimal change to get VkScript to work and we can update to make AmberScript work.

For AmberScript each command RUN, etc, has the pipeline attached so the engine will know what state needs to be set to run a given command. I'm going to fake that in VkScript as well and attach the fake pipeline to each command, just in a future CL.